node/cn/peer.go

```diff
@@ -965,30 +965,33 @@ func (p *multiChannelPeer) UpdateRWImplementationVersion() {
 
 func (p *multiChannelPeer) ReadMsg(rw p2p.MsgReadWriter, connectionOrder int, errCh chan<- error, wg *sync.WaitGroup, closed <-chan struct{}) {
 	defer wg.Done()
-	for {
-		msg, err := rw.ReadMsg()
-		if err != nil {
-			p.GetP2PPeer().Log().Warn("ProtocolManager failed to read msg", "err", err)
-			errCh <- err
-			return
-		}
 
-		msgCh, err := p.chMgr.GetChannelWithMsgCode(connectionOrder, msg.Code)
-		if err != nil {
-			p.GetP2PPeer().Log().Warn("ProtocolManager failed to get msg channel", "err", err)
-			errCh <- err
-			return
-		}
-
-		if msg.Size > ProtocolMaxMsgSize {
-			err := errResp(ErrMsgTooLarge, "%v > %v", msg.Size, ProtocolMaxMsgSize)
-			p.GetP2PPeer().Log().Warn("ProtocolManager over max msg size", "err", err)
-			errCh <- err
-			return
+	readMsgCh := make(chan p2p.Msg, channelSizePerPeer)
+	go func() {
+		for {
+			msg, err := rw.ReadMsg()
+			if err != nil {
+				p.GetP2PPeer().Log().Warn("ProtocolManager failed to read msg", "err", err)
+				errCh <- err
+				return
+			}
+			readMsgCh <- msg
 		}
+	}()
 
+	for {
 		select {
-		case msgCh <- msg:
+		case msg := <-readMsgCh:
+			msgCh, err := p.chMgr.GetChannelWithMsgCode(connectionOrder, msg.Code)
+			if err != nil {
+				p.GetP2PPeer().Log().Warn("ProtocolManager failed to get msg channel", "err", err)
+				errCh <- err
+			} else if msg.Size > ProtocolMaxMsgSize {
+				err := errResp(ErrMsgTooLarge, "%v > %v", msg.Size, ProtocolMaxMsgSize)
+				p.GetP2PPeer().Log().Warn("ProtocolManager over max msg size", "err", err)
+				errCh <- err
+			}
+			msgCh <- msg
```

Comment:
Is is okay to send msg to msgCh when `err != nil` ?

Reply:
Thanks, `return` is ommitted. I added at here https://github.com/klaytn/klaytn/pull/1389/commits/e0c0d77c53127e786fdb20b172d1db60f3a9b74b.

Reply:
When `msgCh` is blocked, it cannot reach `case <-closed:` line. Is it okay?

Reply:
@aidan-kwon @yoomee1313 How about this?
This way, we keep the original `select case msgCh <- msg`.

```go
for {
    var msg Message
    var err error
    select {
    case pair := <-readMsgCh:
        msg, err = pair.Msg, pair.Error
    case <-closed:
        return
    }
    
    if err != nil { ... return } 

    select {
    case msgCh <- msg:
    case <-closed:
        return
    }
}
```

Reply:
@aidan-kwon yes, as you said, it can be blocked..
@blukat29 it helps a lot! It solves the problem, thank you. https://github.com/klaytn/klaytn/pull/1389/commits/8161c47eb91fbb02cf4df6fc6977ade028b65c45.